module/zfs/zfs_vnops.c

```diff
@@ -3079,6 +3166,26 @@ zfs_setattr(struct inode *ip, vattr_t *vap, int flags, cred_t *cr)
 	 * updated as a side-effect of calling this function.
 	 */
 
+	if (projid != ZFS_INVALID_PROJID && !(zp->z_pflags & ZFS_PROJID)) {
+		if (attrzp)
+			err = sa_add_projid(attrzp->z_sa_hdl, tx, projid);
+		if (err == 0)
+			err = sa_add_projid(zp->z_sa_hdl, tx, projid);
+
+		if (unlikely(err == EEXIST)) {
+			err = 0;
+		} else if (err != 0) {
+			goto out;
+		} else {
+			projid = ZFS_INVALID_PROJID;
+			if (xva_req_single(xvap, XAT_PROJID) &&
+			    xva_req_empty(tmpxvattr) && mask == ATTR_XVATTR) {
```

Comment:
Setting the variable 'projid' as "ZFS_INVALID_PROJID" will prevent being repeatedly processed by subsequent logic, such as line 3204, line 3289.

The logic here is that: if the XAT_PROJID is unique valid one in the xvap bitmap, means the caller only needs to change project ID, then after updating the project ID attribute, we can go to 'out' directly since all done. Otherwise, if there are other attributes to be updated, then go ahead, since we have set the variable 'projid' as "ZFS_INVALID_PROJID", it will not be repeatedly handled by subsequent project logic.